feat: add global.images registry, pullSecrets, and pullPolicy support to newrelic-infra-operator

dpacheconr opened the pull request:
## Summary

Add support for cascading registry, pullSecrets, and pullPolicy from global settings to newrelic-infra-operator, following the standard NR chart precedence model: **chart-specific → global → chart default**.

### Phase 1: Global.Images.Registry Support ✅
- Added `registry:` field to all image configs (`image:`, `admissionWebhooksPatchJob.image:`, `config.infraAgentInjection.agentConfig.image:`) to allow chart-specific overrides
- Added `newrelic-infra-operator.sidecar.image` helper to resolve sidecar registry with correct precedence
- Updated configmap to use the new image helper for sidecar injection

### Phase 2: Global.Images.PullSecrets Support ✅
- Added global.images.pullSecrets support to deployment and other manifests
- Global and chart-level pullSecrets are concatenated together

### Phase 3: Global.Images.PullPolicy Support ✅
- Added pullPolicy helpers to `_helpers.tpl`:
  - `newrelic-infra-operator.imagePullPolicy` for operator image
  - `newrelic-infra-operator.admissionWebhooksPatchJob.imagePullPolicy` for webhook patch job image
  - `newrelic-infra-operator.sidecar.imagePullPolicy` for injected sidecar image
- Updated relevant manifests to use pullPolicy helpers
- Nulled out `pullPolicy: IfNotPresent` defaults in values.yaml so global can take effect when no chart-level policy is explicitly set

## Configuration Hierarchy

**For Image Registry:**
1. Chart-level `image.registry` / `admissionWebhooksPatchJob.image.registry` / `config.infraAgentInjection.agentConfig.image.registry`
2. `global.images.registry`
3. Chart default (e.g. `registry.k8s.io` for the webhook job image)

**For ImagePullSecrets:**
1. `global.images.pullSecrets` and chart-level `image.pullSecrets` are **concatenated** (both applied)

**For ImagePullPolicy:**
1. Chart-level `image.pullPolicy` / `admissionWebhooksPatchJob.image.pullPolicy` / `config.infraAgentInjection.agentConfig.image.pullPolicy`
2. `global.images.pullPolicy`
3. Default: `IfNotPresent`

## Backward Compatibility

✅ **Fully backward compatible:**
- Existing values.yaml configurations continue to work unchanged
- Chart-level registry/pullPolicy settings always override global

## Test Plan

**Registry:**
- ✓ Global registry is used when no chart-level registry is set: `my.registry.io/newrelic/infrastructure-k8s:...`
- ✓ Defaults to standard image when neither chart nor global registry is configured
- ✓ Chart-level registry overrides global: `chart.registry.io/newrelic/infrastructure-k8s:...`

**PullSecrets:**
- ✓ No pullSecrets set → No imagePullSecrets section rendered
- ✓ Global pullSecrets only → Uses global
- ✓ Both global + chart pullSecrets → Uses both (concatenated)

**PullPolicy:**
- ✓ No overrides → defaults to `IfNotPresent`
- ✓ Global pullPolicy set → all images adopt it
- ✓ Chart-level pullPolicy overrides global

## Impact

Users can now configure image pull behavior globally:

```yaml
global:
  images:
    registry: "my.private.registry.com"
    pullSecrets:
      - name: my-registry-credentials
    pullPolicy: Always
```

Chart-level overrides still take precedence:

```yaml
image:
  registry: "override.registry.com"   # wins over global
  pullPolicy: Never                   # wins over global
```

This is part of a broader effort to ensure all New Relic Helm charts in the nri-bundle support consistent private registry configuration.

dpacheconr committed:
feat: support global.images.registry for operator and sidecar images

Make the newrelic-infra-operator images respect the global.images.registry
setting, enabling consistent private registry configuration across nri-bundle.

Changes:
- Updated _helpers.tpl with global.images.registry support
- Makes the infrastructure-k8s sidecar image respect global registry settings
- Updated configmap to use the new image helper for sidecar injection

Implementation details:
- If global.images.registry is set AND repository matches default, uses global registry
- If repository is explicitly set to custom value, that takes precedence
- If global registry is not set, defaults to standard paths

Tested scenarios:
✓ Global registry is used when set and repository is default
✓ Defaults to standard image paths when global registry not configured
✓ Explicit repository configuration overrides global registry

dpacheconr committed:
feat: add global.images.pullSecrets support across all templates

Update imagePullSecrets handling to cascade from global.images.pullSecrets:
- Added helpers for both operator and admission webhooks patch job images
- Global and chart-level pullSecrets are concatenated together

Configuration hierarchy:
1. global.images.pullSecrets (highest priority)
2. Chart-level image.pullSecrets
Both sources are concatenated to support flexible secret management.

dpacheconr committed:
feat: add global.images.pullPolicy support to newrelic-infra-operator

Add support for cascading pullPolicy from global.images.pullPolicy setting:
- Added pullPolicy helpers to _helpers.tpl for operator and webhook patch job
- Updated manifests to use pullPolicy helpers

Configuration hierarchy for pullPolicy:
1. global.images.pullPolicy (highest priority)
2. Chart-level image.pullPolicy
3. Default: IfNotPresent

Users can now configure pull policy globally:
  global:
    images:
      pullPolicy: Always

kondracek-nr committed:
chore: fix nil pointer error

jamescripter commented on charts/newrelic-infra-operator/templates/_helpers.tpl:
```diff
@@ -128,9 +128,65 @@ infraAgentInjection:
+{{- if and $registry (eq $imageRepository $defaultRepository) -}}
+  {{- printf "%s/%s" $registry $defaultRepository -}}
+{{- else -}}
+  {{- $imageRepository -}}
```
why not include `$registry` here?

jamescripter commented on charts/newrelic-infra-operator/templates/_helpers.tpl:
```diff
@@ -128,9 +128,65 @@ infraAgentInjection:
+{{- if $globalPullPolicy -}}
+  {{- $globalPullPolicy -}}
+{{- else if $chartPullPolicy -}}
+  {{- $chartPullPolicy -}}
```
why prefer global policy over chart policy?

jamescripter commented on charts/newrelic-infra-operator/templates/_helpers.tpl:
```diff
@@ -128,9 +128,65 @@ infraAgentInjection:
+{{- if $globalPullPolicy -}}
+  {{- $globalPullPolicy -}}
+{{- else if $chartPullPolicy -}}
+  {{- $chartPullPolicy -}}
```
why prefer global policy over chart policy?

dpacheconr committed:
fix: correct value precedence in global image helpers

Subchart-specific values now take precedence over global values, matching
the agreed order: subchart-specific > global > subchart default.

- pullPolicy helpers: chart-level policy wins over global.images.pullPolicy
- sidecar image: global.images.registry applies to any repository, not only
  the default newrelic/infrastructure-k8s

jamescripter commented on charts/newrelic-infra-operator/templates/_helpers.tpl:
```diff
@@ -128,9 +128,65 @@ infraAgentInjection:
+*/}}
+{{- define "newrelic-infra-operator.sidecar.image" -}}
+{{- $imageRepository := .Values.config.infraAgentInjection.agentConfig.image.repository -}}
+{{- $defaultRepository := "newrelic/infrastructure-k8s" -}}
```
I don't think that this is used anymore

dpacheconr committed:
fix: enforce chart-specific → global → default precedence for registry, pullPolicy

- Add registry: field to all image configs (image:, admissionWebhooksPatchJob.image:,
  config.infraAgentInjection.agentConfig.image:) for chart-specific overrides
- Null out pullPolicy defaults in values.yaml so global.images.pullPolicy can take
  effect when no chart-level policy is explicitly set
- Add sidecar.imagePullPolicy helper applying chart → global → IfNotPresent precedence
- Apply sidecar pullPolicy in configmap.data alongside registry override
- Fix double-prefix bug by using deepCopy in configmap.data to avoid mutating .Values

dpacheconr committed:
style(tests): add YAML anchors to global inheritance test suite

Define &base anchor (licenseKey, cluster) on first eligible test and
apply <<: *base across 32 tests. Cluster inheritance tests remain
explicit (no local cluster set). Also resolve CHANGELOG rebase conflict.

dpacheconr committed:
Merge branch 'newrelic:main' into support/global-images-registry

dpacheconr committed:
docs: add helm-docs comments to all registry fields in values.yaml

Add consistent `# --` doc comments to all three registry fields
explaining the precedence model (chart-specific → global → default).

jamescripter committed:
do not render registry since it is not needed internally
